feat: Add graph diagram plugin and related aws-arch/cdk-graph improvements

JeremyJonas opened the pull request:
Plugin to generate diagrams (dot, svg, png) from cdk graph.

JeremyJonas committed:
fix(aws-arch): extract RDS instance types and impl getter

JeremyJonas committed:
fix(aws-arch): improve asset mappings

JeremyJonas committed:
test(cdk-graph): add rds fixture to test

JeremyJonas committed:
feat(aws-arch): dark theme support and minor improvements

JeremyJonas committed:
feat(cdk-graph): graph improvements and fixes

- add `WRAPPED_CFN_RESOURCE` flag and improve handling of wrapped nodes
- add `max` option to `Node.findAncestor`
- change filter "root" terminology to "focus"

JeremyJonas committed:
feat(cdk-graph-plugin-diagram): initial commit

JeremyJonas committed:
feat(cdk-graph): add common filters and improve root handling

JeremyJonas committed:
feat(cdk-graph): add common filters and improve existing filters

- refactored presets into filters and wrapped filters in presets
- created include/exclude filters for both cfnTypes and nodeTypes
- created uncluster filter to remove containers

JeremyJonas committed:
test(cdk-graph-plugin-diagram): add filter tests and update readme examples

JeremyJonas committed:
fix(cdk-graph-plugin-diagram): use roboto open-source font

JeremyJonas committed:
fix(cdk-graph-plugin-diagram): use relative asset paths in dot

JeremyJonas committed:
test(cdk-graph-plugin-diagram): increase image snapshot threshold for ci and  customize name

JeremyJonas committed:
fix(cdk-graph-plugin-diagram): update workspace deps

agdimech commented on packages/aws-arch/src/arch.ts:
```diff
@@ -756,33 +789,80 @@ export class AwsArchitecture {
       const themedIcon = `${qualifiedAssetKey}.${theme}.${format}`;
       if (AwsAsset.AssetFiles.has(themedIcon)) {
         return themedIcon;
+      } else {
```
Why do we need this empty else?

agdimech commented on packages/cdk-graph-plugin-diagram/README.md:
```diff
@@ -0,0 +1,335 @@
+[![API Documentation](https://img.shields.io/badge/view-API_Documentation-blue.svg)](https://aws.github.io/aws-prototyping-sdk/typescript/cdk-graph-plugin-diagram/index.html)
+[![Source Code](https://img.shields.io/badge/view-Source_Code-blue.svg)](https://github.com/aws/aws-prototyping-sdk/tree/mainline/packages/cdk-graph-plugin-diagram)
+
+This plugins generates diagrams utilizing the [cdk-graph](https://aws.github.io/aws-prototyping-sdk/typescript/cdk-graph/index.html) framework.
```
plugin*

agdimech commented on packages/cdk-graph-plugin-diagram/README.md:
````diff
@@ -0,0 +1,335 @@
+### Quick Start
+```ts
+// bin/app.ts
+(async () => {
````
We should add some more docs regarding the async wrapper and that people will need to ensure this surrounds their apps.

agdimech commented on packages/cdk-graph-plugin-diagram/README.md:
````diff
@@ -0,0 +1,335 @@
+// => cdk.out/diagram.png
+```
+
+> This plugin current only support `async report()` generation follow the above example code.
````
This sentence needs work.

agdimech commented on packages/cdk-graph-plugin-diagram/scripts/patch-ts-graphvis.ts:
```diff
@@ -0,0 +1,49 @@
+      tsGraphvizPkg: `${tsGraphvizPkg.name}@${tsGraphvizPkg.version}`,
+    })
+
+    await execa.command("npx downlevel-dts . ts3.9 --to=3.9", {
```
noice :)

agdimech commented on packages/cdk-graph-plugin-diagram/src/internal/graphviz/diagram.ts:
```diff
@@ -0,0 +1,299 @@
+import { GraphTheme, GraphThemeConfigProp } from "./theme";
+
+/**
+ * Union of entities support by diagram
```
supported*

agdimech commented on packages/cdk-graph-plugin-diagram/src/internal/graphviz/diagram.ts:
```diff
@@ -0,0 +1,299 @@
+        entity = new Diagram.StackCluster(gNode as Graph.StackNode);
+        break;
+      }
+      case NodeTypeEnum.DEFAULT: {
```
Can't we include this as part of the below default block?

agdimech commented on packages/cdk-graph-plugin-diagram/scripts/patch-ts-graphvis.ts:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env ts-node
```
I don't see ts-node as a devDep

JeremyJonas committed:
fix(cdk-graph-diagram-plugin): batch of minor diagram improvements

- add sequence diagram for plugin artifact flow
- minor gramar and code fixes
- bundle improvement for low-level lodash pacakges

agdimech committed:
Merge branch 'mainline' into feature/diagram

agdimech reviewed (approved):
Thanks for adding the diagram - much easier to follow!
